### PR TITLE
python v2 plugin: consistent linking for interpreter

### PR DESCRIPTION
**Pull request by cjp256**
When creating a venv, the venv's bin/python[3|3.x] will
always point to the python that created it.  There are
three basic cases where python points to:

(1) the base @ /usr/bin/<python> (typical).

(2) $SNAPCRAFT_PART_INSTALL/usr/bin/<python> (staged in the part).

(3) $SNAPCRAFT_STAGE/usr/bin/<python> (staged from another part).

The first two cases were handled previously, when PATH matched
expected environment.  The third case may occur when another part
stages prior to the python part.

- Resolve the interpreter before creating the venv.  This way it
  is apparent which python the venv wil point to, and provide
  a hint in case things go wrong (e.g. partial staging of python).
  Use this resolved path for the shebang find/replace.

- Fix the broken case where python is resolved to $SNAPCRAFT_STAGE.
  Handle it similar to $SNAPCRAFT_PART_INSTALL, making a relative
  symlink when found.

- While here, refactor some shell code into functions with logging
(set -x) disabled to reduce logging clutter.

Signed-off-by: Chris Patterson <chris.patterson@canonical.com>

### CODE REVIEW

File: snapcraft/plugins/v2/python.py [hunk 1]
```diff
@@ -162,19 +169,25 @@ def get_build_commands(self) -> List[str]:
...
-                    new_link=$(realpath \\
-                               --strip \\
-                               --relative-to="${SNAPCRAFT_PART_INSTALL}/bin/" \\
-                              "${current_link}")
-                    rm "${interp_path}"
-                    ln -s "${new_link}" "${interp_path}"
-                fi
-            fi
+            determine_link_target() {
+                set +x
```
Comment (cjp256) [inline snapcraft/plugins/v2/python.py:173]:
```suggestion
                state="$(set +x)"
```

File: snapcraft/plugins/v2/python.py [hunk 2]
```diff
@@ -162,19 +169,25 @@ def get_build_commands(self) -> List[str]:
...
+                target="${SNAPCRAFT_PYTHON_PATH}"
+                for dir in "${SNAPCRAFT_PART_INSTALL}" "${SNAPCRAFT_STAGE}"; do
+                    if  echo "${SNAPCRAFT_PYTHON_PATH}" | grep -q "${dir}"; then
+                        target="$(realpath --strip --relative-to="${interp_dir}" \\
+                                "${SNAPCRAFT_PYTHON_PATH}")"
+                        break
+                    fi
+                done
+                echo "${target}"
+                set -x
```
Comment (cjp256) [inline snapcraft/plugins/v2/python.py:184]:
```suggestion
                eval "$(state)"
```

File: tests/unit/plugins/v2/test_python.py [hunk 3]
```diff
@@ -67,32 +67,34 @@ def test_get_build_environment():
...
+                target="${SNAPCRAFT_PYTHON_PATH}"
+                for dir in "${SNAPCRAFT_PART_INSTALL}" "${SNAPCRAFT_STAGE}"; do
+                    if  echo "${SNAPCRAFT_PYTHON_PATH}" | grep -q "${dir}"; then
+                        target="$(realpath --strip --relative-to="${interp_dir}" \\
+                                "${SNAPCRAFT_PYTHON_PATH}")"
+                        break
+                    fi
+                done
+                echo "${target}"
+                set -x
```
Comment (cjp256) [inline tests/unit/plugins/v2/test_python.py:90]:
```suggestion
               eval "$(opts_state)"
```

File: tests/unit/plugins/v2/test_python.py [hunk 4]
```diff
@@ -67,32 +67,34 @@ def test_get_build_environment():
...
-                           --strip \\
-                           --relative-to="${SNAPCRAFT_PART_INSTALL}/bin/" \\
-                          "${current_link}")
-                rm "${interp_path}"
-                ln -s "${new_link}" "${interp_path}"
-            fi
-        fi
-    """
+            determine_link_target() {
+                set +x
```
Comment (cjp256) [inline tests/unit/plugins/v2/test_python.py:79]:
```suggestion
                opts_state="$(set +x)"
```

File: tests/unit/plugins/v2/test_python.py [hunk 5]
```diff
@@ -67,32 +67,34 @@ def test_get_build_environment():
...
+                target="${SNAPCRAFT_PYTHON_PATH}"
+                for dir in "${SNAPCRAFT_PART_INSTALL}" "${SNAPCRAFT_STAGE}"; do
+                    if  echo "${SNAPCRAFT_PYTHON_PATH}" | grep -q "${dir}"; then
+                        target="$(realpath --strip --relative-to="${interp_dir}" \\
+                                "${SNAPCRAFT_PYTHON_PATH}")"
+                        break
+                    fi
+                done
+                echo "${target}"
+               eval "$(opts_state)"
```
Comment (cjp256) [inline tests/unit/plugins/v2/test_python.py:90]:
```suggestion
                eval "$(opts_state)"
```

File: tests/unit/plugins/v2/test_python.py [hunk 6]
```diff
@@ -67,32 +67,34 @@ def test_get_build_environment():
...
-                           --strip \\
-                           --relative-to="${SNAPCRAFT_PART_INSTALL}/bin/" \\
-                          "${current_link}")
-                rm "${interp_path}"
-                ln -s "${new_link}" "${interp_path}"
-            fi
-        fi
-    """
+            determine_link_target() {
+                opts_state="$(set +x)"
```
Comment (cjp256) [inline tests/unit/plugins/v2/test_python.py:79]:
```suggestion
                opts_state="$(set +o +x)"
```

File: snapcraft/plugins/v2/python.py [hunk 7]
```diff
@@ -162,19 +169,25 @@ def get_build_commands(self) -> List[str]:
...
-                    new_link=$(realpath \\
-                               --strip \\
-                               --relative-to="${SNAPCRAFT_PART_INSTALL}/bin/" \\
-                              "${current_link}")
-                    rm "${interp_path}"
-                    ln -s "${new_link}" "${interp_path}"
-                fi
-            fi
+            determine_link_target() {
+                opts_state="$(set +x)"
```
Comment (cjp256) [inline snapcraft/plugins/v2/python.py:173]:
```suggestion
                opts_state="$(set +o +x)"
```

**Review by cmatsuoka — APPROVED** [review]
It took me a while to understand what was happening there but in the end it seems to make sense. It could be interesting to have a comment in the code with a few more details on the covered scenarios (maybe something similar to what's explained in the PR description).

**Review by sergiusens — APPROVED** [review]

**Pull request merged**